admission-webhook/src/utils/policy-manager.ts

```diff
@@ -0,0 +1,139 @@
+    this._policyInformer.informer.on('update', this.onPolicy.bind(this));
+    this._policyInformer.informer.on('delete', this.onPolicyRemoval.bind(this));
+
+    this._bindingInformer.informer.on('add', this.onBinding.bind(this));
+    this._bindingInformer.informer.on('update', this.onBinding.bind(this));
+    this._bindingInformer.informer.on('delete', this.onBindingRemoval.bind(this));
+  }
+
+  async start() {
+    await this._policyInformer.start();
+    await this._bindingInformer.start();
+  }
+
+  getMatchingPolicies(resource: AdmissionRequestObject, namespace: string): MonokleApplicablePolicy[] {
+    this._logger.debug({policies: this._policies.size, bindings: this._bindings.size});
+
+    if (this._bindings.size === 0) {
+      return [];
+    }
+
+    return Array.from(this._bindings.values())
+      .map((binding) => {
+        const policy = this._policies.get(binding.spec.policyName);
+
+        if (!policy) {
+          this._logger.error({msg: 'Binding is pointing to missing policy', binding});
+          return null;
+        }
+
+        if (binding.spec.matchResources && !this.isResourceMatching(binding, resource)) {
+          return null;
+        }
+
+        return {
+          policy: policy.spec,
+          binding: binding.spec
+        }
+      })
+      .filter((policy) => policy !== null) as MonokleApplicablePolicy[];
+  }
+
+  private onPolicy(policy: MonoklePolicy) {
+    this._logger.debug({msg: 'Policy updated', policy});
+
+    this._policies.set(policy.metadata!.name!, policy);
+
+    this.emit('policyUpdated', policy);
+  }
+
+  private onPolicyRemoval(policy: MonoklePolicy) {
+    this._logger.debug({msg: 'Policy removed', policy});
+
+    this._policies.delete(policy.metadata!.name!);
+
+    this.emit('policyRemoved', policy);
+  }
+
+  private onBinding(binding: MonoklePolicyBinding) {
+    this._logger.debug({msg: 'Binding updated', binding});
+
+    this._bindings.set(binding.metadata!.name!, binding);
+
+    this.emit('bindingUpdated', binding);
+  }
+
+  private onBindingRemoval(binding: MonoklePolicyBinding) {
+    this._logger.debug({msg: 'Binding removed', binding});
+
+    this._bindings.delete(binding.metadata!.name!);
+
+    this.emit('bindingRemoved', binding);
+  }
+
+  private isResourceMatching(binding: MonoklePolicyBinding, resource: AdmissionRequestObject): boolean {
+    const namespaceMatchLabels = binding.spec.matchResources?.namespaceSelector?.matchLabels;
+
+    this._logger.trace({
+      msg: 'Checking if resource matches binding',
+      namespaceMatchLabels,
+      resourceMetadata: resource.metadata.labels
+    });
+
+    if (!namespaceMatchLabels) {
+      return true;
+    }
+
+    for (const key of Object.keys(namespaceMatchLabels)) {
+      if (resource.metadata.labels?.[key] !== namespaceMatchLabels[key]) {
+        if (!(key === 'namespace' && resource.metadata.namespace === namespaceMatchLabels[key])) {
```

Comment:
See [this comment](https://github.com/kubeshop/monokle-admission-controller/pull/1/files#r1343612624). This is a bit "stretched" interpretation of namespace matching (when resource doesn't have `metadata.label.namespace`). But I have some doubts about it.

Reply:
The link is broken. Not sure I understand the problem correctly. For normal Kubernetes Objects you would look to `metadata.namespace` because `metadata.label.namespace` will rarely exist.

Reply:
The link lead to my previous comment which you already answered with:

> This is such a common use case that Kubernetes automatically adds a label to all namespaces called `kubernetes.io/metadata.name=$namespaceName`.

So my understanding is that below matching which we want to support:

```
  matchResources:
    namespaceSelector:
      matchLabels:
        namespace: default
```

is `metadata.label` matcher according to specs. Meaning being compliant with specs, it should only try to match `resource.metadata.label.namespace`?

Or maybe not 🤔 It doesn't really say how matching is done for Namespaced resources:

> NamespaceSelector decides whether to run the admission control policy on an object based on whether the namespace for that object matches the selector. If the object itself is a namespace, the matching is performed on object.metadata.labels. If the object is another cluster scoped resource, it never skips the policy.

Not sure now TBH 🤔

Reply:
It's a 2-way lookup:
- Match me all the resources that match this namespace selector.
- Give me the namespace that match these labels.

You can see it in action in the excerpt below from [this blogpost](https://medium.com/google-cloud/validating-admission-policies-with-gke-1-26-ed1321bcf739). 

```
const namespaces = matchExpressions(allnamespaces, ...); // Returns all namespaces but these system namespaces
const objects = matchResources(namespaces); // Returns all namespaces looking at `meta.namespace`
```

```
apiVersion: admissionregistration.k8s.io/v1beta1
kind: ValidatingAdmissionPolicyBinding
metadata:
  name: max-replicas-deployments
spec:
  paramRef:
    name: max-replicas-deployments
    namespace: policies-configs
    parameterNotFoundAction: Deny
  policyName: max-replicas-deployments
  validationActions:
    - Deny
  matchResources:
    namespaceSelector:
      matchExpressions:
      - key: kubernetes.io/metadata.name
        operator: NotIn
        values:
        - kube-node-lease
        - kube-public
        - kube-system
        - allow-listed
```

Reply:
@WitoDelnat thanks for the explanation, now I understand👍 Well, not it works fine with matching resources in given namespace (thanks to checking `metadata.namespace` as you described) and selector matching. So it seems to be consistent with what you described.

For more general handling of `matchResources` with `matchExpressions`, `resourceRules` and others I will add an issue in `monokle-core` as you suggested 👍

Reply:
Reported as https://github.com/kubeshop/monokle-core/issues/538.